### PR TITLE
RUN-1665: Add trait utilities to help identify non-deterministic iteration order

### PR DESCRIPTION
**Pull request by sgeisenh**
https://linear.app/replay/issue/RUN-1665/detect-and-automatically-make-all-instances-of-wtf-set-and-map

Start of some static analysis utilities for identifying cases of iteration over containers that may have different iteration orders between runs.

### CODE REVIEW

File: third_party/blink/renderer/platform/wtf/type_traits.h [hunk 1]
```diff
@@ -104,6 +106,45 @@ struct IsSubclassOfTemplateTypenameSizeTypename {
...
+struct IsPointerType : std::false_type {};
+
+template <typename T>
+struct IsPointerType<T*> : std::true_type {};
+
+template <typename T>
+struct IsPointerType<std::unique_ptr<T>> : std::true_type {};
+
+template <typename T>
+struct IsPointerType<scoped_refptr<T>> : std::true_type {};
```
Comment (Domiii) [inline third_party/blink/renderer/platform/wtf/type_traits.h:129]: A few more pointer types I can think of. Not sure if they matter all too much:

* `Persistent`
* `WeakPersistent`
* `WeakPtr` / `WeakPtrBase`
* `raw_ptr`

**Review by Domiii — COMMENTED** [review]
Looks good to me! But I'll defer the review to those with a more solid understanding of template magic 🙄

File: third_party/blink/renderer/platform/heap/member.h [hunk 2]
```diff
@@ -133,8 +133,9 @@ struct MemberHashRecordReplayRegisteredPointerId
   static bool Equal(const U& a, const V& b) {
     return a == b;
   }
-};
 
+  static constexpr bool IsDeterministicHash = true;
```
Comment (bhackett1024) [inline third_party/blink/renderer/platform/heap/member.h:137]: I think IsRecordReplayDeterministicHash would be better, this makes it easy to see this code is a recording related change we've made when browsing the source.

File: third_party/blink/renderer/platform/wtf/type_traits.h [hunk 3]
```diff
@@ -104,6 +106,45 @@ struct IsSubclassOfTemplateTypenameSizeTypename {
...
+//
+// Example usage:
+//   inline typename HashSet<T, U, V, W>::iterator HashSet<T, U, V, W>::begin()
+//       const {
+//     static_assert(!IsProblematicContainerV<T, U>,
+//                   "Using non-deterministic pointer hash");
+//     return impl_.begin();
+//   }
+template <typename Key, typename Hash>
+constexpr bool IsProblematicContainerV =
```
Comment (bhackett1024) [inline third_party/blink/renderer/platform/wtf/type_traits.h:145]: Similarly, IsRecordReplayNonDeterministicContainerV ?

**Review by bhackett1024 — APPROVED** [review]

File: third_party/blink/renderer/platform/heap/member.h (hunk 2 above)
Comment (sgeisenh) [reply third_party/blink/renderer/platform/heap/member.h]: That makes sense. Done.

File: third_party/blink/renderer/platform/wtf/type_traits.h (hunk 3 above)
Comment (sgeisenh) [reply third_party/blink/renderer/platform/wtf/type_traits.h]: Done.

File: third_party/blink/renderer/platform/heap/member.h [hunk 4]
```diff
@@ -161,7 +162,7 @@ struct MemberHashRecordReplayId
           return Base::GetHash(st.GetAsInteger());
         }
       } else {
-        CHECK(id > 0); // Should have been registered.
+        CHECK(id > 0);  // Should have been registered.
```
[the remark this answers is not in the document]
Comment (sgeisenh) [reply third_party/blink/renderer/platform/heap/member.h]: Removed.

File: third_party/blink/renderer/platform/wtf/type_traits.h (hunk 1 above)
Comment (sgeisenh) [reply third_party/blink/renderer/platform/wtf/type_traits.h:136]: Thank you! Added.

**Pull request merged**